core/chains/evm/txm/txm.go

```diff
@@ -0,0 +1,443 @@
+	defer cancel()
+	broadcastWithBackoff := newBackoff(1 * time.Second)
+	var broadcastCh <-chan time.Time
+
+	for {
+		start := time.Now()
+		bo, err := t.broadcastTransaction(ctx, address)
+		if err != nil {
+			t.lggr.Errorw("Error during transaction broadcasting", "err", err)
+		} else {
+			t.lggr.Debug("Transaction broadcasting time elapsed: ", time.Since(start))
+		}
+		if bo {
+			broadcastCh = time.After(broadcastWithBackoff.Duration())
+		} else {
+			broadcastWithBackoff.Reset()
+			broadcastCh = time.After(utils.WithJitter(broadcastInterval))
+		}
+		select {
+		case <-ctx.Done():
+			return
+		case <-triggerCh:
+			continue
+		case <-broadcastCh:
+			continue
+		}
+	}
+}
+
+func (t *Txm) backfillLoop(address common.Address) {
+	defer t.wg.Done()
+	ctx, cancel := t.stopCh.NewCtx()
+	defer cancel()
+	backfillWithBackoff := newBackoff(t.config.BlockTime)
+	backfillCh := time.After(utils.WithJitter(t.config.BlockTime))
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-backfillCh:
+			start := time.Now()
+			bo, err := t.backfillTransactions(ctx, address)
+			if err != nil {
+				t.lggr.Errorw("Error during backfill", "err", err)
+			} else {
+				t.lggr.Debug("Backfill time elapsed: ", time.Since(start))
+			}
+			if bo {
+				backfillCh = time.After(backfillWithBackoff.Duration())
+			} else {
+				backfillWithBackoff.Reset()
+				backfillCh = time.After(utils.WithJitter(t.config.BlockTime))
+			}
+		}
+	}
+}
+
+func (t *Txm) broadcastTransaction(ctx context.Context, address common.Address) (bool, error) {
+	for {
+		_, unconfirmedCount, err := t.txStore.FetchUnconfirmedTransactionAtNonceWithCount(ctx, 0, address)
+		if err != nil {
+			return false, err
+		}
+
+		// Optimistically send up to 1/3 of the maxInFlightTransactions. After that threshold, broadcast more cautiously
+		// by checking the pending nonce so no more than maxInFlightTransactions/3 can get stuck simultaneously i.e. due
+		// to insufficient balance. We're making this trade-off to avoid storing stuck transactions and making unnecessary
+		// RPC calls. The upper limit is always maxInFlightTransactions regardless of the pending nonce.
+		if unconfirmedCount >= maxInFlightTransactions/maxInFlightSubset {
+			if unconfirmedCount > maxInFlightTransactions {
+				t.lggr.Warnf("Reached transaction limit: %d for unconfirmed transactions", maxInFlightTransactions)
+				return true, nil
+			}
+			pendingNonce, e := t.client.PendingNonceAt(ctx, address)
+			if e != nil {
+				return false, e
+			}
+			nonce := t.getNonce(address)
+			if nonce > pendingNonce {
+				t.lggr.Warnf("Reached transaction limit. LocalNonce: %d, PendingNonce %d, unconfirmedCount: %d",
+					nonce, pendingNonce, unconfirmedCount)
+				return true, nil
+			}
+		}
+
+		tx, err := t.txStore.UpdateUnstartedTransactionWithNonce(ctx, address, t.getNonce(address))
+		if err != nil {
+			return false, err
+		}
+		if tx == nil {
+			return false, nil
+		}
+		nonce := t.getNonce(address)
```

Comment:
nit: couldn't we just do `t.getNonce(address)` once above `UpdateUnstartedTransactionWithNonce`?

Reply:
Actually doesn't `UpdateUnstartedTransactionWithNonce` return a tx with `Nonce` and `State` already set?

Reply:
IIRC I tried to distinguish the Storage updates from the local copy updates so it's more obvious to the reader what's going on and in case we had a DB implementation I would first update the DB state and then the local state. But I think, since the `UpdateUnstartedTransactionWithNonce` is self-explanatory I can simplify this.